File: ios/Reguerta/Reguerta/Data/Calendar/InMemoryDeliveryCalendarRepository.swift
import Foundation

actor InMemoryDeliveryCalendarRepository: DeliveryCalendarRepository {
    private let defaultDay: DeliveryWeekday
    private var overrides: [String: DeliveryCalendarOverride] = [:]

    init(defaultDay: DeliveryWeekday = .wednesday) {
        self.defaultDay = defaultDay
    }

    func defaultDeliveryDayOfWeek(environment _: SessionEnvironment) async -> DeliveryWeekday { defaultDay }

    func allOverrides(environment _: SessionEnvironment) async -> [DeliveryCalendarOverride] {
        overrides.values.sorted { $0.weekKey < $1.weekKey }
    }

    func upsertOverride(
        _ override: DeliveryCalendarOverride,
        environment _: SessionEnvironment
    ) async -> DeliveryCalendarOverride {
        overrides[override.weekKey] = override
        return override
    }

    func deleteOverride(weekKey: String, environment _: SessionEnvironment) async {
        overrides.removeValue(forKey: weekKey)
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/calendar/InMemoryDeliveryCalendarRepository.kt
package com.reguerta.user.data.calendar

import com.reguerta.user.domain.calendar.DeliveryCalendarOverride
import com.reguerta.user.domain.calendar.DeliveryCalendarRepository
import com.reguerta.user.domain.calendar.DeliveryWeekday
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class InMemoryDeliveryCalendarRepository(
    private val defaultDeliveryDayOfWeek: DeliveryWeekday = DeliveryWeekday.WEDNESDAY,
) : DeliveryCalendarRepository {
    private val mutex = Mutex()
    private val overrides = linkedMapOf<String, DeliveryCalendarOverride>()

    override suspend fun getDefaultDeliveryDayOfWeek(): DeliveryWeekday = defaultDeliveryDayOfWeek

    override suspend fun getAllOverrides(): List<DeliveryCalendarOverride> = mutex.withLock {
        overrides.values.sortedBy { it.weekKey }
    }

    override suspend fun upsertOverride(override: DeliveryCalendarOverride): DeliveryCalendarOverride = mutex.withLock {
        overrides[override.weekKey] = override
        override
    }

    override suspend fun deleteOverride(weekKey: String) {
        mutex.withLock {
            overrides.remove(weekKey)
        }
    }
}
